### PR TITLE
test_framework: detect failure of bitcoind startup

### PR DESCRIPTION
**Pull request by laanwj**
Replace the `bitcoin-cli -rpcwait` after spawning bitcoind with our own loop that detects when bitcoind exits prematurely (before RPC is up).

This prevents a hang in such a case (see #7463).

### CODE REVIEW

**Comment by laanwj** [comment]
Not ready to merge yet. I've tested this with a -salvagewallet loop:
```
Unexpected exception caught during testing: bitcoind exited with status 1 during initialization
  File "/data/src/bitcoin/qa/rpc-tests/test_framework/test_framework.py", line 135, in main
    self.run_test()
  File "./wallet.py", line 295, in run_test
    self.nodes = start_nodes(3, self.options.tmpdir, [[m]] * 3)
  File "/data/src/bitcoin/qa/rpc-tests/test_framework/util.py", line 283, in start_nodes
    return [ start_node(i, dirname, extra_args[i], rpchost, binary=binary[i]) for i in range(num_nodes) ]
  File "/data/src/bitcoin/qa/rpc-tests/test_framework/util.py", line 267, in start_node
    wait_for_bitcoind_start(bitcoind_processes[i], url, i)
  File "/data/src/bitcoin/qa/rpc-tests/test_framework/util.py", line 144, in wait_for_bitcoind_start
    raise Exception('bitcoind exited with status %i during initialization' % process.returncode)
Stopping nodes
```
Even though reporting is better, it still hangs after "Stopping nodes", going to investigate why.

**Comment by laanwj** [comment]
Should be fixed now. The problem was that start_nodes can fail with part of the nodes already started.
`wait_bitcoinds` will subsequently wait forever for them to exit.
One-to-last commit makes sure that they're stopped properly in that case.

**Comment by laanwj** [comment]
This is faster than the current -rpcwait construction, so #7745 triggers more, seems I need to solve that too here.

File: qa/rpc-tests/test_framework/util.py [hunk 1]
```diff
@@ -267,7 +280,14 @@ def start_nodes(num_nodes, dirname, extra_args=None, rpchost=None, binary=None):
     """
     if extra_args is None: extra_args = [ None for i in range(num_nodes) ]
     if binary is None: binary = [ None for i in range(num_nodes) ]
-    return [ start_node(i, dirname, extra_args[i], rpchost, binary=binary[i]) for i in range(num_nodes) ]
+    rpcs = []
+    try:
+        for i in range(num_nodes):
+            rpcs.append(start_node(i, dirname, extra_args[i], rpchost, binary=binary[i]))
+    except: # If one node failed to start, stop the others
```
Comment (MarcoFalke) [inline qa/rpc-tests/test_framework/util.py]: Nice!

**Comment by jonasschnelli** [comment]
Good idea!
utACK 018b60c5ea703ed12edcde034a185f79e77e5576

**Comment by MarcoFalke** [comment]
utACK 018b60c

File: qa/rpc-tests/test_framework/util.py [hunk 2]
```diff
@@ -130,11 +131,33 @@ def initialize_datadir(dirname, n):
         f.write("listenonion=0\n")
     return datadir
 
+def rpc_url(i, rpchost=None):
+    return "http://rt:rt@%s:%d" % (rpchost or '127.0.0.1', rpc_port(i))
```
Comment (MarcoFalke) [inline qa/rpc-tests/test_framework/util.py]: Nit: `rpcbind_test.py` no longer works after this.

Comment (laanwj) [inline qa/rpc-tests/test_framework/util.py]: We should probably re-add that test to one of the standard sequence, otherwise it will just bitrot

**Pull request merged**